# INF2542 • PEMBELAJARAN MESIN
## Pertemuan 06: Python for Machine Learning
**Topik:** Dari Data Mentah → Data Bersih → Informasi Visual → Fitur Siap Dimodelkan  
**Pustaka Utama:** Google Colab / Jupyter • NumPy • Pandas • Matplotlib • Scikit-Learn

---
### Identitas Mahasiswa:
- **Nama:** Gathan Hilabi
- **NIM:** 60324059
- **Mata Kuliah:** Machine Learning (INF2542)
- **Pertemuan:** 06 (Fondasi Coding & Data Preparation Pipeline)
- **Alokasi Waktu:** Blok Pertemuan 06–07 (300 Menit)
- **Sub-CPMK 9.1:** Terampil menggunakan pustaka Python dan menggunakannya untuk membangun proses pengolahan data ML.

---
### Workflow Praktikum Hari Ini:
$$\text{1. Load (Baca Dataset)} \longrightarrow \text{2. Inspect (Cek Kualitas)} \longrightarrow \text{3. Clean (Perbaiki Masalah)} \longrightarrow \text{4. Transform (Bentuk Fitur)} \longrightarrow \text{5. Visualize (Cari Pola)} \longrightarrow \text{6. Prepare (Siapkan X dan y)}$$

> **Prinsip Utama:** *Jangan membersihkan data tanpa tahu masalah apa yang sedang diperbaiki!*  
> Data cleaning bukan tahap terpisah dari ML — ia adalah bagian fundamental dari pipeline machine learning.

---
## 1. NumPy - Fondasi Data Numerik (Slide 06)

### Konsep & Intuisi:
NumPy (*Numerical Python*) menyediakan struktur data array $N$-dimensi (`ndarray`) dan operasi matematika tervektorisasi yang sangat efisien. Model Machine Learning bekerja pada representasi numerik yang terstruktur, di mana data observasi diatur sebagai baris dan fitur sebagai kolom.

Pada contoh ini, membuat matriks fitur $X$ berukuran $3 \times 3$:
- **Baris (Observasi):** 3 observasi mahasiswa
- **Kolom (Fitur):** Kehadiran (%), IPK, dan Jam Belajar (jam/minggu)
- **Parameter `axis=0`:** Operasi reduksi/agregasi dilakukan secara vertikal menyusuri baris (per kolom/fitur).

In [1]:
import numpy as np

X = np.array([
    [90, 3.70, 8],
    [75, 3.20, 5],
    [88, 3.60, 7]
])

print("Shape:", X.shape)
print("Mean :", X.mean(axis=0))
print("Min  :", X.min(axis=0))
print("Max  :", X.max(axis=0))

Shape: (3, 3)
Mean : [84.33333333  3.5         6.66666667]
Min  : [75.   3.2  5. ]
Max  : [90.   3.7  8. ]


### Interpretasi NumPy (Slide 06):
1. **Shape `(3, 3)`:** Menunjukkan matriks berdimensi 2 dengan 3 baris (observasi mahasiswa) dan 3 kolom (fitur numerik).
2. **Operasi `axis=0`:** Menghitung nilai statistik agregasi per fitur:
   - **Rata-rata (Mean):** Kehadiran = 84.33%, IPK = 3.50, Jam Belajar = 6.67 jam.
   - **Nilai Minimum (Min):** Kehadiran = 75%, IPK = 3.20, Jam Belajar = 5 jam.
   - **Nilai Maksimum (Max):** Kehadiran = 90%, IPK = 3.70, Jam Belajar = 8 jam.
3. **Mengapa Penting?** NumPy adalah fondasi dasar dari seluruh ekosistem komputasi saintifik Python. Library tingkat tinggi seperti Pandas dan Scikit-Learn secara internal mengonversi data tabel ke dalam struktur array NumPy untuk melakukan komputasi aljabar linier dan optimasi parameter model.

---
## 2. Praktikum Terbimbing - Satu Notebook, Satu Alur (Slide 07 – 18)

### Studi Kasus: Data Mahasiswa (`data_mahasiswa.csv`)
Mengolah dataset mahasiswa mentah yang sengaja mengandung beberapa masalah kualitas data:
- Missing value (nilai kosong)
- Format angka tidak konsisten
- Duplikasi baris observasi
- Kategori status tidak konsisten

Target akhir: Mengubah data mentah menjadi data bersih, menghasilkan visualisasi eksploratif, serta memisahkan fitur $X$ dan target $y$ yang siap digunakan untuk model Scikit-Learn.